src/cpio/client_providers/private_key_fetcher_provider/azure/azure_private_key_fetcher_provider_utils.cc

```diff
@@ -26,19 +27,438 @@ using google::scp::azure::attestation::hasSnp;
 using google::scp::core::HttpMethod;
 using google::scp::core::HttpRequest;
 using google::scp::core::Uri;
+using google::scp::cpio::client_providers::AzurePrivateKeyFetchingClientUtils;
 
 namespace google::scp::cpio::client_providers {
 
+bool AzurePrivateKeyFetchingClientUtils::isPrivate(EVP_PKEY* pkey) {
+  // Determine if the key is private or public
+  int key_type = EVP_PKEY_type(pkey->type);
+  bool is_private = false;
+  if (key_type == EVP_PKEY_RSA) {
+    const RSA* rsa_key = EVP_PKEY_get1_RSA(pkey);
```

Comment:
As we discussed in the previous PR, we need to read the manual carefully for each line of OpenSSL code...
```
EVP_PKEY_get1_RSA(), EVP_PKEY_get1_DSA(), EVP_PKEY_get1_DH() and EVP_PKEY_get1_EC_KEY() return the referenced key in pkey or NULL if the key is not of the correct type. The returned key must be freed after use. 
```
https://www.openssl.org/docs/man3.3/man3/EVP_PKEY_get1_RSA.html

So this is a memory leak in this example. Can you create a issue to check and fix these problem.
I haven't checked all the OpenSSL usage, but we should because its spec is not obvious usually.

We need to check mainly things like:
- We should free memory
- If it's caller, we need to use the wrapper to make sure it's always freed
- Check error
- Free the error queue if the document requires